design/one-pager-template-stacks.md

```diff
@@ -0,0 +1,344 @@
+
+In this example, the `.spec.version` value would be taken from the body of each CR this Stack manages. That value would be applied to the template and rendered as a new or updated `Deployment` resource.
+
+Should the `controller.deployment` value be treated as a template? This could have unintended consequences for existing Stack deployments, especially if their definitions includes pieces that match the template format.
+
+To avoid disturbing existing `controller.deployment` values, let's use a new template aware `templates` spec field. We'll take that approach in the following sections.
+
+### No Custom Controller Needed
+
+Existing Stacks must specify a `Deployment`. These are typically written in Go and generally require understanding some combination of client-go, controller-runtime, crossplane-runtime, and Kubebuilder.
+
+For some applications, the intended work of the controller could be simple:
+
+* For every new resource, create a set of Kubernetes resources
+* Use the spec fields of that resource to affect the Kubernetes resources
+* Upon changes to these spec fields, update the Kubernetes resources
+
+This is where the value of Template Stacks starts to materialize.
+
+```yaml
+kind: Stack
+metadata:
+  name: RedisStack
+spec:
+  templates:
+    redis.example.org/v1:
+      deployment: |
+          …
+          containers:
+          - name: redis-controller
+            image: example/redis-controller:{{.spec.redisVersion}}
+```
+
+*In this example, templates are nested within a map keyed by the GVR and a template grouping name. This will be explained in the [One Pager for Template Stack Fields](one-pager-template-stack-fields).*
+
+### Composition
+
+#### Vertical Composition
+
+Crossplane Stacks manage multiple CRDs. Template Stacks should be no different.
+
+Template definitions will need to express the template to render for each CRD that is managed. The GVR can be used as a map key to denote the CRD template being described.
+
+Each template needed to render a GVR should be provided (demonstrated below as a map), additionally a name that represents that intent of that template should distinguish each template (also demonstrated as a map). This name can then be reused to reference a template (see [Horizontal Composition](#Horizontal-Composition)).
+
+```yaml
+kind: Stack
+metadata:
+  name: CachingWebServiceStack
+spec:
+  templates:
+    redis.example.org/v1:
+      deployment: |
+        kind: Deployment
+        …
+          containers:
+          - name: redis-controller
+            image: example/redis-controller:{{.spec.redisVersion}}
+    cachingwebservice.example.org/v1:
+      cache: |
+        apiVersion: example.org/v1
+        kind: Redis
+        …
+        spec:
+          redisVersion: "5"
+      web: |
+        kind: Deployment
+        …
+          containers:
+          - name: nginx-controller
+            image: example/nginx-controller:{{.spec.nginxVersion}}
+--
```

Comment:
I've been trying to be consistent about the use of `…`.

If I can avoid writing more stock YAML between a first level and a nest value, I'll put the ellipsis in between.

Since the order of resource values is not important, an ellipsis on the top or the bottom of a resource carries equal value, but I will tend to start with the ellipsis so it is more obvious that I am skipping over some values.

This may have been more of a convenience for me than the reader.  I was trying to draw focus to the important pieces without the distraction of other elements.

Would it be better if I included valid YAML blocks in all examples?

Reply:
I think I see what you mean here.. I didn't include the '...' on `CachingWebService`.. Github inline messages sometime leave me wonder what line the comment was concerned with.  I'll add the '...'.

Reply:
I went with `apiVersion` instead here.. Good call.